custom-notification-handler/src/main/java/org/wso2/carbon/sample/CustomNotificationHandler.java

```diff
@@ -0,0 +1,191 @@
+ * added to the buildNotification() method in the CustomNotificationUtil class).
+ * <p>
+ * Changing the email template type during the user self signup REST API.
+ * ---------------------------------------------------------------------
+ * Default email template type will be overridden if the self signup request parameter map contains the
+ * customTemplateType key with the value of the custom template.
+ * <p>
+ * <p>
+ * <p>
+ * Role based email template change during the user password reset
+ * ----------------------------------------------------------
+ * <p>
+ * Default email template will be overridden by checking a specific role of the user. A registry resource
+ * with the path /identity/config/customRoleBasedEmailTemplateTypeConfig needs to be configured by
+ * providing a set of registry properties mapping a role to a custom email template
+ * as follows.
+ * <p>
+ * propertyName= <role name>  propertyValue= <custom email template type>
+ * <p>
+ * e.g. propertyName = role1  propertyValue = PasswordResetForRole1
+ * <p>
+ * If the user is assigned to one of the roles configured as above in the registry, the corresponding custom
+ * email template will be used.
+ */
+public class CustomNotificationHandler extends NotificationHandler {
+
+    private static final String CLAIM_URI_ROLE = "http://wso2.org/claims/role";
+    private static final String CUSTOM_TEMPLATE_TYPE_PROPERTY_KEY = "customTemplateType";
+    private static final String ACCOUNT_CONFIRMATION_EMAIL_TEMPLATE_TYPE = "AccountConfirmation";
+    private static final String PASSWORD_RESET_EMAIL_TEMPLATE_TYPE = "PasswordReset";
+    private static final String CONFIG_LOCATION = "/identity/config/customRoleBasedEmailTemplateTypeConfig";
+
+    private static Log log = LogFactory.getLog(CustomNotificationHandler.class);
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        Map<String, String> placeHolderData = new HashMap<>();
+
+        for (Map.Entry<String, Object> entry : event.getEventProperties().entrySet()) {
+            if (entry.getValue() instanceof String) {
+                placeHolderData.put(entry.getKey(), (String) entry.getValue());
+            }
+        }
+
+        String notificationEvent = (String) event.getEventProperties().get(NotificationConstants.EmailNotification
+                .EMAIL_TEMPLATE_TYPE);
+
+        if (ACCOUNT_CONFIRMATION_EMAIL_TEMPLATE_TYPE.equalsIgnoreCase(notificationEvent) && placeHolderData.containsKey
+                (CUSTOM_TEMPLATE_TYPE_PROPERTY_KEY)) {
+            event.getEventProperties().put(NotificationConstants.EmailNotification
+                    .EMAIL_TEMPLATE_TYPE, placeHolderData.get(CUSTOM_TEMPLATE_TYPE_PROPERTY_KEY));
+            if (log.isDebugEnabled()) {
+                log.debug("Using the custom email template type " + placeHolderData.get(CUSTOM_TEMPLATE_TYPE_PROPERTY_KEY));
```

Comment:
`"Using the custom email template type " `-> `"Using the custom email template type: "` to differentiate the property value